### PR TITLE
[WIP] Eliminate some heap allocation of tuples in ilread.fs

### PR DESCRIPTION
**Pull request by TIHan**
After doing a bit of profiling when opening VisualFSharp.sln, there are significant allocations due to tuples, specifically in ilread.fs. I believe almost all of them are shorted lived and gen0.

untitled

In the image, the column `Exc %`, is the percentage across the entirety of VS for allocations when opening VisualFSharp.sln for this sample. Sometimes when profiling we will see `System.String` be very high in percentage, mainly due to the project system; but in this case, the project system wasn't doing too much work for this sample, allowing us to see the rest of the percentages. 

This PR is intended to eliminate these allocations on the heap so GC isn't being bombarded. Instead, we need to allocate on the stack and pass by reference to not copy anything.

There are multiple approaches to solving this problem, so I expect opinions. :)

### CODE REVIEW

File: src/absil/ilread.fs [hunk 1]
```diff
@@ -1264,14 +1347,20 @@ let seekReadConstantRowUncached ctxtH idx =
...
+
 /// Read Table CustomAttribute.
 let seekReadCustomAttributeRow (ctxt: ILMetadataReader)  idx =
     let mdv = ctxt.mdfile.GetView()
     let mutable addr = ctxt.rowAddr TableNames.CustomAttribute idx
     let parentIdx = seekReadHasCustomAttributeIdx ctxt mdv &addr
     let typeIdx = seekReadCustomAttributeTypeIdx ctxt mdv &addr
     let valIdx = seekReadBlobIdx ctxt mdv &addr
-    (parentIdx, typeIdx, valIdx)  
+    { ParentIndex = parentIdx; TypeIndex = typeIdx; ValueIndex = valIdx }
```
Comment (TIHan) [inline src/absil/ilread.fs]: I'm wondering if this is a good case for anonymous records. :)

**Comment by TIHan** [comment]
The first commit eliminates allocations from the top percentage of tuples from the image, `3.5%`.

**Comment by dsyme** [comment]
@TIHan The code looks fine on a quick glance - the approach is fine. But there must be something causing the test failure.

**Comment by TIHan** [comment]
It is quite hard to measure this change, as the perf benefits are quite small, but my best guess is that we get, for example, 100ms back when compiling FSharp.Core.

Been testing this quite a bit. Changing to a value type did render lower gen0 collections and the number of allocations on the heap. With the latest change "Improved CPU time slightly", this just allocates one struct and passes it byref and gets its fields set; this actually improves CPU time.

I do think its worth having lower gen0 collections, so I'm still in favor of being on the stack. With the latest change where I allocate the struct once and pass it byref, this isn't quite idiomatic but does improve CPU time. I was able to measure the total sum of time of all calls to `seekReadCustomAttributeRow` when compiling FSharp.Core:

**Allocate struct each time - these numbers are nearly the same if it was not a struct as well**
```
367.68ms
353.53ms
365.29ms
```

**Allocate struct once, set fields**
```
329.90ms
321.60ms
324.38ms
```

**Comment by TIHan** [comment]
Something to note:

```
let seekReadCustomAttributeRow (ctxt: ILMetadataReader)  idx (attrRow: byref<CustomAttributeRow>) =
    let mdv = ctxt.mdfile.GetView()
    let mutable addr = ctxt.rowAddr TableNames.CustomAttribute idx
    attrRow.parentIndex <- seekReadHasCustomAttributeIdx ctxt mdv &addr
    attrRow.typeIndex <- seekReadCustomAttributeTypeIdx ctxt mdv &addr
    attrRow.valueIndex <- seekReadBlobIdx ctxt mdv &addr
```
and
```
let seekReadCustomAttributeRow (ctxt: ILMetadataReader)  idx (attrRow: byref<CustomAttributeRow>) =
    let mdv = ctxt.mdfile.GetView()
    let mutable addr = ctxt.rowAddr TableNames.CustomAttribute idx
    let parentIdx = seekReadHasCustomAttributeIdx ctxt mdv &addr
    let typeIdx = seekReadCustomAttributeTypeIdx ctxt mdv &addr
    let valIdx <- seekReadBlobIdx ctxt mdv &addr

    attrRow.parentIndex <- parentIdx
    attrRow.typeIndex <- typeIdx
    attrRow.valueIndex <- valIdx
```

The latter is measurably slower than the former. I wondering if we could eliminate the let binding.

**Comment by TIHan** [comment]
Yes, I intend to finish.

**Comment by KevinRansom** [comment]
@TIHan do you still intend to finish this?

Kevin

**Comment by TIHan** [comment]
Not sure, since I did a prototype of a re write of ilread.fs it gets rid of these allocations as well

**Comment by dsyme** [comment]
@TIHan I think the prototype you did is really strong, so I'd say we can drop this since the scale of improvement is very different?

**Comment by cartermp** [comment]
Agreed since the rewrite is just so much better on initial benchmarking

**Comment by TIHan** [comment]
Closing as we will have a better solution in the future.

**Comment by TIHan** [comment]
To be more specific, I have a prototype of a re-implementation of `ilread.fs` using SRM (System.Reflection.Metadata): https://github.com/TIHan/visualfsharp/blob/new-metadata-reader/src/fsharp/FSharp.Compiler.Private/ilread_srm.fs. `ilread.fsi` is the same API, just its implementation detail is entirely SRM.

**Pull request closed without merging**